Add EIP: Block-scoped transient storage opcodes (#11467)
ethereum/EIPs: EIPS/eip-8212.md

description: Add opcodes for manipulating state that behaves identically to storage but is discarded after every block
author: monokh (@monokh)
discussions-to: https://ethereum-magicians.org/t/eip8212-block-scoped-transient-storage/28137
status: Draft
type: Standards Track
category: Core
created: 2026-03-30
requires: 1153, 2200
---

## Abstract

This proposal introduces block-scoped transient storage opcodes, which manipulate state that behaves identically to storage, except that block-scoped transient storage is discarded after every block. Unlike transaction-scoped transient storage introduced in [EIP-1153](./eip-1153.md), values written via `BSTORE` persist across transactions within the same block and are readable by subsequent transactions executing against the same contract. All values are discarded when the block is finalized. Because block-scoped transient storage is never serialized to disk, it is significantly cheaper than persistent storage (`SSTORE`/`SLOAD`). Block-scoped transient storage is accessible to smart contracts via 2 new opcodes, `BLOAD` and `BSTORE`, where "B" stands for "block:"

```
BLOAD  (0x5e)
BSTORE (0x5f)
```

## Motivation

> [EIP-1153](./eip-1153.md) introduced transaction-scoped transient storage, enabling cheap inter-frame communication within a single transaction. However, there is no equivalent mechanism for cheap state that persists across transactions within the same block. Currently, contracts that need to maintain block-level state — such as cumulative values, per-block rate limits, or block-local registries — must use persistent storage (`SSTORE`/`SLOAD`), paying the full cost of disk serialization for data that is only meaningful for the duration of a single block.

Reviewer: ```suggestion
[EIP-1153](./eip-1153.md) introduced transaction-scoped transient storage, enabling cheap inter-frame communication within a single transaction. However, there is no equivalent mechanism for cheap state that persists across transactions within the same block. Currently, contracts that need to maintain block-level state — such as cumulative values, per-block rate limits, or block-local registries — must use persistent storage (`SSTORE (0x55)`/`SLOAD (0x54)`), paying the full cost of disk serialization for data that is only meaningful for the duration of a single block.
```

(see https://eips.ethereum.org/EIPS/eip-1#special-requirements-for-core-eips)

This is wasteful. Persistent storage incurs high gas costs because it must be written to and read from disk, replicated across all nodes, and maintained indefinitely in the state trie. For state that is inherently ephemeral to a single block, these costs are disproportionate to the value of the data.

Block-scoped transient storage solves this by providing a dedicated storage region that:

1. **Never touches disk.** Values exist only in memory for the lifetime of the block, so reads and writes are dramatically cheaper than persistent storage.

> 2. **Persists across transactions within a block.** Unlike `TSTORE`/`TLOAD` which are cleared after each transaction, `BSTORE`/`BLOAD` values accumulate across all transactions in a block, enabling new patterns of cross-transaction coordination.

Reviewer: ```suggestion
2. **Persists across transactions within a block.** Unlike `TSTORE (0x5d)`/`TLOAD (0x5c)` which are cleared after each transaction, `BSTORE`/`BLOAD` values accumulate across all transactions in a block, enabling new patterns of cross-transaction coordination.
```

3. **Block-scoped registries:** Temporary registries or caches that are meaningful only within the current block context.
4. **Cheaper state transitions for block-basis operations:** Any contract logic that resets or rotates state on a per-block basis can use block-scoped storage instead of persistent storage, significantly reducing gas costs.
5. **MEV-aware contract logic:** Contracts can observe and react to cumulative activity within a block, enabling more sophisticated on-chain strategies.
6. **Cross-transaction reentrancy guards:** Locks that span an entire block rather than a single transaction.

These opcodes are more efficient to execute than `SSTORE` and `SLOAD` because values are never serialized to or deserialized from disk. They are also more efficient than writing to persistent storage and later overwriting, because no storage refund accounting is needed.

## Specification

Two new opcodes are added to the EVM, `BLOAD` (`0x5e`) and `BSTORE` (`0x5f`).

They use the same arguments on stack as `SLOAD` (`0x54`) and `SSTORE` (`0x55`).

> `BLOAD` pops one 32-byte word from the top of the stack, treats this value as the address, fetches a 32-byte word from the block-scoped transient storage at that address for the executing contract, and pushes the value on top of the stack. If no value has been written to that slot in the current block, the value returned is 0.

Reviewer: This is confusing, I believe you mean this pops two stack items `[address, key]` and they reads `key` from `address` and pushes that to stack?

Reviewer: Ah, never mind, I thought it was intended to store a `value` at another `address` (another account (EOA or contract). This is not the case. The "address" is confusing here, I suggest `slot` as you use later in the sentence also.

```suggestion
`BLOAD` pops one 32-byte word `slot` from the top of the stack,  fetches a 32-byte word from the block-scoped transient storage at that `slot` for the executing contract, and pushes the value on top of the stack. If no value has been written to that slot in the current block, the value returned is 0.
```

> `BSTORE` pops two 32-byte words from the top of the stack. The word on the top is the address, and the next is the value. `BSTORE` saves the value at the given address in the block-scoped transient storage for the executing contract.

Reviewer: Also explicitly define this here, like 
```suggestion
`BSTORE` pops two 32-byte words `[slot, value]` from the top of the stack. `BSTORE` saves the `value` at the given `slot` in the block-scoped transient storage for the executing contract.
```

> Addressing is the same as `SLOAD` and `SSTORE`. i.e. each 32-byte address points to a unique 32-byte word.

Reviewer: ```suggestion
Addressing is the same as `SLOAD` and `SSTORE`, i.e. each 32-byte slot points to a unique 32-byte word.
```

I'm not sure if this sentence is necessary though.

> Gas cost for `BSTORE` is the same as a warm `SSTORE` of a dirty slot (i.e. original value is not new value and is not current value, currently 100 gas), and gas cost of `BLOAD` is the same as a hot `SLOAD` (value has been read before, currently 100 gas). Gas cost cannot be on par with memory access due to block-scoped transient storage's interactions with reverts.

Reviewer: Have to check but 

> (i.e. original value is not new value and is not current value, currently 100 gas)

I do not think this is correct (I think it is higher)

Block-scoped transient storage is private to the contract that owns it, in the same way as persistent storage. Only owning contract frames may access their block-scoped transient storage. And when they do, all the frames access the same block-scoped transient store, in the same way as persistent storage, but unlike memory.

When block-scoped transient storage is used in the context of `DELEGATECALL` or `CALLCODE`, then the owning contract of the block-scoped transient storage is the contract that issued `DELEGATECALL` or `CALLCODE` instruction (the caller) as with persistent storage. When block-scoped transient storage is used in the context of `CALL` or `STATICCALL`, then the owning contract of the block-scoped transient storage is the contract that is the target of the `CALL` or `STATICCALL` instruction (the callee).

If a frame reverts, all writes to block-scoped transient storage that took place between entry to the frame and the return are reverted, including those that took place in inner calls. This mimics the behavior of persistent storage and transaction-scoped transient storage as defined in [EIP-1153](./eip-1153.md).

> If the `BSTORE` opcode is called within the context of a `STATICCALL`, it will result in an exception instead of performing the modification. `BLOAD` is allowed within the context of a `STATICCALL`.

Reviewer: Review comment: it depends on how you interpret STATICCALL, it at least means that it does not edit the state trie, which is the case here. For TLOAD/TSTORE it is fine in STATICCALL, but here not. I can see arguments for both sides. Blocking it is the safest way though (so this is fine). But there are context where we need to STATICCALL but still want to BLOAD/BSTORE, and we would likely also want to know if we are in STATICCALL (because then we switch to TSTORE instead of BSTORE (?))
